okhttp/src/main/java/okhttp3/CertificatePinner.kt

```diff
@@ -0,0 +1,339 @@
+            if (pin.hash == sha1) return  // Success!
+          }
+          else -> throw AssertionError("unsupported hashAlgorithm: ${pin.hashAlgorithm}")
+        }
+      }
+    }
+
+    // If we couldn't find a matching pin, format a nice exception.
+    val message = buildString {
+      append("Certificate pinning failure!")
+      append("\n  Peer certificate chain:")
+      for (c in 0 until peerCertificates.size) {
+        val x509Certificate = peerCertificates[c] as X509Certificate
+        append("\n    ")
+        append(pin(x509Certificate))
+        append(": ")
+        append(x509Certificate.subjectDN.name)
+      }
+      append("\n  Pinned certificates for ")
+      append(hostname)
+      append(":")
+      for (pin in pins) {
+        append("\n    ")
+        append(pin)
+      }
+    }
+    throw SSLPeerUnverifiedException(message)
+  }
+
+  @Deprecated(
+      "replaced with {@link #check(String, List)}.",
+      ReplaceWith("check(hostname, peerCertificates.toList())")
+  )
+  @Throws(SSLPeerUnverifiedException::class)
+  inline fun check(hostname: String, vararg peerCertificates: Certificate) {
+    check(hostname, peerCertificates.toList())
+  }
+
+  /**
+   * Returns list of matching certificates' pins for the hostname. Returns an empty list if the
+   * hostname does not have pinned certificates.
+   */
+  internal fun findMatchingPins(hostname: String): List<Pin> {
+    var result: List<Pin> = emptyList()
+    for (pin in pins) {
+      if (pin.matches(hostname)) {
+        if (result.isEmpty()) result = mutableListOf()
+        (result as MutableList<Pin>).add(pin)
+      }
+    }
+    return result
+  }
+
+  /** Returns a certificate pinner that uses `certificateChainCleaner`.  */
+  internal fun withCertificateChainCleaner(
+    certificateChainCleaner: CertificateChainCleaner?
+  ): CertificatePinner {
+    return if (this.certificateChainCleaner == certificateChainCleaner) {
+      this
+    } else {
+      CertificatePinner(pins, certificateChainCleaner)
+    }
+  }
+
+  override fun equals(other: Any?): Boolean {
+    if (other === this) return true
+    return other is CertificatePinner
+        && certificateChainCleaner == other.certificateChainCleaner
+        && pins == other.pins
+  }
+
+  override fun hashCode(): Int {
+    var result = Objects.hashCode(certificateChainCleaner)
+    result = 31 * result + pins.hashCode()
+    return result
+  }
+
+  internal class Pin(
+    /** A hostname like `example.com` or a pattern like `*.example.com`.  */
+    val pattern: String,
+    pin: String
+  ) {
+    /** The canonical hostname, i.e. `EXAMPLE.com` becomes `example.com`.  */
+    val canonicalHostname: String
+    /** Either `sha1/` or `sha256/`.  */
+    val hashAlgorithm: String
+    /** The hash of the pinned certificate using [.hashAlgorithm].  */
+    val hash: ByteString
+
+    init {
+      this.canonicalHostname = if (pattern.startsWith(WILDCARD)) {
```

Comment:
inline into val above?

Reply:
I don’t like it for big initializers but I don’t feel strongly.